src/rp2_common/hardware_rtc/rtc.c

```diff
@@ -114,60 +139,88 @@ void rtc_enable_alarm(void) {
-    if (t->dotw  < 0) return true;
-    if (t->hour  < 0) return true;
-    if (t->min   < 0) return true;
-    if (t->sec   < 0) return true;
-    return false;
+static repeat_type rtc_alarm_repeats(const datetime_t *t) {
+  // If any value is set to -1 then we don't match on that value
+  // hence the alarm will eventually repeat
+    if (t->year < 0 && t->month < 0 && t->day < 0 && t->dotw < 0
+        && t->hour < 0 && t->min < 0 && t->sec < 0) return CONTINUOUS_REPEAT_EVERY_SEC;
+
+    return (t->year < 0 || t->month < 0 || t->day < 0 || t->dotw < 0
+            || t->hour < 0 || t->min < 0 || t->sec < 0)
+        ? CONTINUOUS_REPEAT : NO_REPEAT;
 }
 
-void rtc_set_alarm(datetime_t *t, rtc_callback_t user_callback) {
-    rtc_disable_alarm();
+bool rtc_set_alarm(const datetime_t *t, rtc_callback_t user_callback) {
+    if (!rtc_running())
+      return false;
 
-    // Only add to setup if it isn't -1
-    rtc_hw->irq_setup_0 = ((t->year  < 0) ? 0 : (((uint)t->year)  << RTC_IRQ_SETUP_0_YEAR_LSB )) |
-                          ((t->month < 0) ? 0 : (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB)) |
-                          ((t->day   < 0) ? 0 : (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB  ));
-    rtc_hw->irq_setup_1 = ((t->dotw  < 0) ? 0 : (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB)) |
-                          ((t->hour  < 0) ? 0 : (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB)) |
-                          ((t->min   < 0) ? 0 : (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB )) |
-                          ((t->sec   < 0) ? 0 : (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB ));
+    rtc_disable_alarm();
 
-    // Set the match enable bits for things we care about
-    if (t->year  >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_YEAR_ENA_BITS);
-    if (t->month >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_MONTH_ENA_BITS);
-    if (t->day   >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_DAY_ENA_BITS);
-    if (t->dotw  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_DOTW_ENA_BITS);
-    if (t->hour  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_HOUR_ENA_BITS);
-    if (t->min   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_MIN_ENA_BITS);
-    if (t->sec   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_SEC_ENA_BITS);
+    uint32_t s0 = 0, s1 = 0;
 
     // Does it repeat? I.e. do we not match on any of the bits
     _alarm_repeats = rtc_alarm_repeats(t);
 
+    bool check_params = (is_valid_datetime(t) || _alarm_repeats != NO_REPEAT) && user_callback;
+    valid_params_if(RTC, check_params);
+    if(!check_params)	// none of the parameters is valid
+        return false;
+
+    // Set the match enable bits for things we care about
+    if(_alarm_repeats == CONTINUOUS_REPEAT_EVERY_SEC) {
+        // repeatable every second! All entries are -1
+        datetime_t new_dt;
+        _rtc_get_datetime(&new_dt);
+        _seconds_increment = (-t->sec) % 60;
```

Comment:
even more confused, because t->sec is -1 here always

Reply:
maybe best described with an example:
if sec input parameter for rtc_set_alarm() is in range 0<=s<=59, than we have absolute behavior. The alarm always appears when internal comparison of s with the RTC seconds is true.

If sec value is negative, it is interpreted as a relative value. Means if it is -1, _seconds_increment is set to 1, and the alarm appears whenever the seconds changes. If now input sec is -2, the increment will be 2 and therefore the ISR is called every 2 seconds. As this input value is just local within rtc_set_alarm(), it must be remembered as a state, so that ISR can access it.

Reply:
E.g. like snooze function on a clock. Set alarm to -30sec repeat, alarm appears, you press button to switch off the alarm, and after 30s it appears again.

Reply:
Oh maybe you're confused because _seconds_increment is processed within the macro ADD_AND_ENABLE_REPEATABLE_SECOND?

Reply:
I think @Reneg973 was intending that you could set e.g. "repeat every 5 seconds" by setting `t->sec` to -5 ?
But I'm not sure if that's desirable behaviour, since e.g. there's no similar functionality to set "repeat every 5 minutes" (and we're not trying to re-implement `cron` :wink: ) so I think it might be more consistent to revert the "repeat every 5 seconds" functionality, and just stick to the "setting everything to -1 causes the alarm to repeat every second" behaviour. For more sophisticated behaviour, the user can always add their own custom logic inside their callback function.

Reply:
no i was confused because I didn't notice you could now set this to negative values other than -1. I'm not sure that this functionality belongs here; there are plenty of other ways of getting repeating timers. `hardware_foo` libraries are generally intended for exposing hardware functionality in a clean API, not using the hardware to provide additional functionality.
I need to take a closer look at the RTC h/w, so I'll get back to you

Reply:
From SW side you are right, this functionality could be implemented in a level above.
There are 2 solutions then
1. set an alarm "repeat every second" -> count the number of IRQs arrived :: fast ISR but highly increased number of IRQs
2. set an alarm at current_time + x -> get the IRQ -> query datetime_t -> increment by x -> set new alarm via rtc_set_alarm() :: slow ISR but not more IRQs than required.

So both methods produce more overhead than really required.
Anyway I don't know which HAL strategy you want to follow for Pico.

Reply:
3rd approach: repeat every second -> query datetime_t -> if `t->sec % 5 == 0` do something, else ignore :wink: 

As I mentioned before, with RP2040 running with a CPU clock of 125MHz I don't think you really need to worry about "slow ISR"?

Reply:
You do need to worry!

Reply:
3rd approach = 1st :) - I meant the same! And t->sec % 5 is not correct, you need include the start seconds.
And 125MHz is just the default. I have improved my SPI performance by 22% (increase SPI freq 31.25MHz -> 40MHz), but for this I had to sacrifice 36% of clk_sys by setting it to 80MHz!

Reply:
> no i was confused because I didn't notice you could now set this to negative values other than -1. I'm not sure that this functionality belongs here; there are plenty of other ways of getting repeating timers. `hardware_foo` libraries are generally intended for exposing hardware functionality in a clean API, not using the hardware to provide additional functionality.
> I need to take a closer look at the RTC h/w, so I'll get back to you

@kilograham:
Repeating timers are not repeating clock ticks. Anyway here is my proposal for a modification to be able to do this outside:
1. generate a new interface function rtc_modify_alarm() which reduces the overhead of rtc_set_alarm() and therefore can be called within the ISR.
2. in almost all use cases when the user ISR is called, the RTC time will be queried. Give it as a parameter (tell don't ask principle).

What do you think?